Handles restarting problem of adaptive time integration methods

ArseniyKholod opened the pull request:
If time integration method uses time step controller like PID controller, than to obtain an exact solution after restarting the computation we have to save some additional data in SaveRestartCallback. This is a first shot.

ArseniyKholod commented:
I wrote a small test for new Callback solving` CompressibleEulerEquations2D` with `SaveRestartCallback` (solver is `SSPRK43()`). 
**Source code with SaveRestartCallback**:
```
using OrdinaryDiffEq
using Trixi

gamma = 1.4
equations = CompressibleEulerEquations2D(gamma)

initial_condition = initial_condition_density_wave

solver = DGSEM(polydeg=5, surface_flux=flux_central)

coordinates_min = (-1.0, -1.0)
coordinates_max = ( 1.0,  1.0)
mesh = TreeMesh(coordinates_min, coordinates_max,
                initial_refinement_level=2,
                n_cells_max=30_000)


semi = SemidiscretizationHyperbolic(mesh, equations, initial_condition, solver)

tspan = (0.0, 1.0)
ode = semidiscretize(semi, tspan);

summary_callback = SummaryCallback()

analysis_interval = 100
analysis_callback = AnalysisCallback(semi, interval=analysis_interval,
                                     extra_analysis_integrals=(entropy, energy_total))

alive_callback = AliveCallback(analysis_interval=analysis_interval)

save_restart = SaveRestartCallback(interval=10,
                                   save_final_restart=true)

save_solution = SaveSolutionCallback(interval=100,
                                     save_initial_solution=true,
                                     save_final_solution=true,
                                     solution_variables=cons2prim)


callbacks = CallbackSet(summary_callback,
                        analysis_callback, alive_callback,
                        save_restart, save_solution)

controller=IController()

sol = solve(ode, SSPRK43(),
            dt=0.001, # solve needs some value here but it will be overwritten by the stepsize_callback
            save_everystep=false, callback=callbacks, controller = controller; ode_default_options()...);

summary_callback()
```
**Source code with new LoadRestartCallback:**
```
using Trixi
using OrdinaryDiffEq

gamma = 1.4
equations = CompressibleEulerEquations2D(gamma)

initial_condition = initial_condition_density_wave

solver = DGSEM(polydeg=5, surface_flux=flux_central)

restart_filename = joinpath("out", "restart_000250.h5")
mesh = load_mesh(restart_filename)

semi = SemidiscretizationHyperbolic(mesh, equations, initial_condition, solver)

tspan = (load_time(restart_filename), 1.0)
ode = semidiscretize(semi, tspan, restart_filename);

summary_callback = SummaryCallback()

analysis_interval = 100
analysis_callback = AnalysisCallback(semi, interval=analysis_interval,
                                     extra_analysis_integrals=(entropy, energy_total))

alive_callback = AliveCallback(analysis_interval=analysis_interval)

load_restart = LoadRestartCallback(restart_file=restart_filename)

save_solution = SaveSolutionCallback(interval=100,
                                     save_initial_solution=true,
                                     save_final_solution=true,
                                     solution_variables=cons2prim)
# load_restart=LoadRestartCallback(restart_filename)

# example_callback = Trixi.LoadRestartCallback(restart_filename)
callbacks = CallbackSet(summary_callback,
                        analysis_callback, alive_callback,
                        load_restart, 
                        save_solution)

controller=IController()

sol = solve(ode, SSPRK43(),
            dt=1.0, # solve needs some value here but it will be overwritten by the stepsize_callback
            save_everystep=false, callback=callbacks, controller = controller; ode_default_options()...);
summary_callback() # print the timer summary
```

**Results are equal for serial and parallel (MPI) running:**
exact PIController 
Variable:       rho              rho_v1           rho_v2           rho_e
 L2 error:       1.28409828e-03   1.28591753e-04   2.56767139e-04   1.11860108e-04
 Linf error:     6.08861719e-03   6.09667406e-04   1.22302535e-03   1.23407359e-03
without new Callback
L2 error:       1.28409972e-03   1.28623381e-04   2.56760373e-04   1.60062668e-04
 Linf error:     6.08853318e-03   6.09652664e-04   1.22447760e-03   1.92716652e-03
with new Callback
L2 error:       1.28409828e-03   1.28591753e-04   2.56767139e-04   1.11860108e-04
 Linf error:     6.08861719e-03   6.09667406e-04   1.22302535e-03   1.23407359e-03

exact PIDController
 L2 error:       1.30237464e-03   1.30567731e-04   2.61033435e-04   8.15273144e-04
 Linf error:     6.04078026e-03   6.30698340e-04   1.20978076e-03   9.62542480e-03
without new Callback
 L2 error:       1.30256493e-03   1.30162163e-04   2.62179813e-04   3.62746252e-04
 Linf error:     6.04201846e-03   6.41746290e-04   1.21186121e-03   5.45822582e-03
with new Callback
 L2 error:       1.30237464e-03   1.30567731e-04   2.61033435e-04   8.15273144e-04
 Linf error:     6.04078026e-03   6.30698340e-04   1.20978076e-03   9.62542480e-03

exact IController
L2 error:       1.28364777e-03   1.28658982e-04   2.56498887e-04   1.29917808e-03
 Linf error:     6.07903127e-03   6.05004458e-04  1.22374874e-03   1.38208249e-02
without new Callback
 L2 error:       1.28364050e-03   1.28181157e-04   2.56900473e-04   3.96760696e-04
 Linf error:     6.07911661e-03   6.10334008e-04   1.21279397e-03   7.20215380e-03
with new Callback
 L2 error:       1.28364777e-03   1.28658982e-04   2.56498887e-04   1.29917808e-03
 Linf error:     6.07903127e-03   6.05004458e-04   1.22374874e-03   1.38208249e-02

So adding a new `LoadRestartCallback ` provides the result exactly as without restarting.

ArseniyKholod commented:
General concept is, that we have to save some data for integrator.controller, to obtain exactly the same solution as without restart. So I added additional functions that saves them in `SaveRestartCallback`. Problem in a restarting file is that we have to load these data after creating an integrator, that occurs inside `solve() ` function. It's the reason why I decided to implement a loading stuff inside new callback `LoadRestartCallback`. While initializing this callback we load all data for integrator.controller. After initializing this callback is never called, because function that activate callback returns always false.

As was said in a discussion in trixi_students I tested that implementation for these solvers:

- SSPRK43()

- RDPK3SpFSAL49()

-all with adaptive stepsize from (Explicit-Strong-Stability-Preserving-Runge-Kutta-Methods-for-Hyperbolic-PDEs)[https://docs.sciml.ai/DiffEqDocs/stable/solvers/ode_solve/#Explicit-Strong-Stability-Preserving-Runge-[…]ods-for-Hyperbolic-PDEs-(Conservation-Laws)](https://docs.sciml.ai/DiffEqDocs/stable/solvers/ode_solve/#Explicit-Strong-Stability-Preserving-Runge-Kutta-Methods-for-Hyperbolic-PDEs-(Conservation-Laws)) and https://docs.sciml.ai/DiffEqDocs/stable/solvers/ode_solve/#Low-Storage-Methods

ArseniyKholod commented:
@SimonCan, what do you think about this idea? 
Of coarse I'll add some asserts a bit later and other checks.

SimonCan commented on Project.toml:
```diff
@@ -19,6 +19,7 @@ MPI = "da04e1cc-30fd-572f-bb4f-1f8673147195"
 MuladdMacro = "46d2c3a1-f734-5fdb-9937-b9b9aeba4221"
 Octavian = "6fd5a793-0b7e-452c-907f-f8bfe9c57db4"
 OffsetArrays = "6fe1bfb0-de20-5000-8ca7-80f57d26f881"
+OrdinaryDiffEq = "1dea7af3-3e70-54e6-95c3-0bf5283fa5ed"
```
Since OrdinaryDiffEq is being loaded in each elixir, you might not need this.

ArseniyKholod replied:
I need it, because I specify different methods for PIController, PIDController. They are not defined without OrdinaryDiffEq package.

SimonCan replied:
Could you then only load the required methods, like:
using OrdinaryDiffEq: PIDController

SimonCan commented on src/Trixi.jl:
```diff
@@ -30,7 +30,7 @@ using Reexport: @reexport
 # as long as HDF5.jl uses Requires.jl to enable parallel HDF5 with MPI
 using MPI: MPI
-
+using OrdinaryDiffEq
```
Same here. Do you need to load OrdinaryDiffEq?

ArseniyKholod replied:
The same like previous

SimonCan commented on src/callbacks_step/load_restart.jl:
Since this is not part of any callback it might make sense to put it in `src/auxiliary` instead.

ArseniyKholod replied:
I defined a new one LoadRestartCallback in it, so it's callback, but only in structure, because it's never called after initialization. So

SimonCan replied:
Maybe we can think in a future PR if we want to change this. We can keep this for the moment.

SimonCan reviewed (changes requested):
Looks great.
Could you add example elixirs like the ones in your comments? I was thinking of asimilar design as in `examples/structured_2d_dgsem/elixir_advection_restart.jl` where they call first call `elixir_advection_extended.jl` and then load the data for a restart.

ArseniyKholod commented:
Hi @SimonCan,
In the last week I re-writed code in a bit more logical way. Nothing new in sense of correctness, but I deleted the new created callback `LoadRestartCallback` and added` load_controller!` function in `SaveRestartCallback` instead, which works in the same way, but does not require additional callback initialization and is consistent with `load_time, load_timestep` and other functions that are called in a restart setup. Also I added test in examples folder, and it is tested in serial, mpi and threaded part of CI. Tests are [examples/tree_2d_dgsem/elixir_euler_density_wave_extended.jl](https://github.com/trixi-framework/Trixi.jl/pull/1565/files#diff-5e21c4a95b358a759a277b7615922397407deb86871e546e1d3d59a98aa55b08) and [examples/tree_2d_dgsem/elixir_euler_density_wave_restart.jl](https://github.com/trixi-framework/Trixi.jl/pull/1565/files#diff-a6ea81e7cc453ffdf6e35f9161c04f59b2a9f08b56d0ec38f4a8c90a744b0fcf), they are based on [tree_2d_dgsem/elixir_euler_density_wave.jl](https://github.com/trixi-framework/Trixi.jl/blob/main/examples/tree_2d_dgsem/elixir_euler_density_wave.jl)

SimonCan commented:
> Hi @SimonCan, In the last week I re-writed code in a bit more logical way. Nothing new in sense of correctness, but I deleted the new created callback `LoadRestartCallback` and added` load_controller!` function in `SaveRestartCallback` instead, which works in the same way, but does not require additional callback initialization and is consistent with `load_time, load_timestep` and other functions that are called in a restart setup. Also I added test in examples folder, and it is tested in serial, mpi and threaded part of CI. Tests are [examples/tree_2d_dgsem/elixir_euler_density_wave_extended.jl](https://github.com/trixi-framework/Trixi.jl/pull/1565/files#diff-5e21c4a95b358a759a277b7615922397407deb86871e546e1d3d59a98aa55b08) and [examples/tree_2d_dgsem/elixir_euler_density_wave_restart.jl](https://github.com/trixi-framework/Trixi.jl/pull/1565/files#diff-a6ea81e7cc453ffdf6e35f9161c04f59b2a9f08b56d0ec38f4a8c90a744b0fcf), they are based on [tree_2d_dgsem/elixir_euler_density_wave.jl](https://github.com/trixi-framework/Trixi.jl/blob/main/examples/tree_2d_dgsem/elixir_euler_density_wave.jl)

That's great. @ArseniyKholod . It is always good to have your code covered by some tests.
I noticed that `load_PIDController` is not covered by your tests. This seems to be because `controller_type == "PID"` is not true for the test examples. To free two birds with one key, you could change your MPI test so that it uses the `PID` controller.

ArseniyKholod commented:
@SimonCan, Thank you for the point! I added exactly that

SimonCan reviewed (commented):
Looks good so far.

SimonCan commented:
You cover a large number of cases. We would need those also in the tests. Currently the function `save_restart_file_parallel` is not covered, because `HDF5.has_parallel()`  is false. For the same reason `load_restart_file_parallel` is not covered by any test. `save_restart_controller(integrator, controller::IController, restart_callback)` is also not covered. For the merge we would need those covered.

SimonCan reviewed (commented):
Looks good.

sloede commented on test/test_tree_2d_euler.jl:
```diff
@@ -26,6 +26,14 @@ EXAMPLES_DIR = pkgdir(Trixi, "examples", "tree_2d_dgsem")
+  @trixi_testset "elixir_euler_density_wave_restart.jl" begin
+    trixi_include(@__MODULE__, joinpath(EXAMPLES_DIR, "elixir_euler_density_wave_extended.jl"))
+    l2, linf = analysis_callback(sol)
+    eval(:(@test_trixi_include(joinpath(EXAMPLES_DIR, "elixir_euler_density_wave_restart.jl"),
+      # Expected errors are exactly the same as in the elixir_euler_density_wave_extended.jl
+      l2 = $l2, linf = $linf)))
```
Why do you need the call to `eval`?

ArseniyKholod replied:
Thank you for a review!
`eval` is used to evaluate macros with runtime arguments, I wanted to prove in runtime, that pause doesn't make any difference in result.

sloede replied:
Are you sure that this is really necessary? I am not convinced, but maybe I am missing something... we did not need this anywhere before, did we?

ArseniyKholod replied:
You're rigth. As I understand macros are normally working only with compilation time known arguments. Before we used `@test_trixi_include` with compilation known `linf` and `l2`, but now I try to use `@test_trixi_include` with only in runtime known linf` and `l2`. It throws an error, if I don't apply any modifications on `@test_trixi_include` . You're rigth, that `eval` is not necessary here, but necessary to use quoting `:` in way `:@test_trixi_include`. I'm not quite sure yet, why `:` helps. But it works somehow...

sloede commented on test/test_threaded.jl:
```diff
@@ -145,6 +145,14 @@ Trixi.mpi_isroot() && isdir(outdir) && rm(outdir, recursive=true)
+      using OrdinaryDiffEq: IController
+      trixi_include(@__MODULE__, joinpath(examples_dir(), "tree_2d_dgsem", "elixir_euler_density_wave_extended.jl"), controller = IController())
+      l2, linf = analysis_callback(sol)
+      eval(:(@test_trixi_include(joinpath(examples_dir(), "tree_2d_dgsem", "elixir_euler_density_wave_restart.jl"), controller = IController(),
```
Why do you need the call to `eval`?

sloede commented on test/test_mpi_tree.jl:
```diff
@@ -189,6 +189,15 @@ CI_ON_WINDOWS = (get(ENV, "GITHUB_ACTIONS", false) == "true") && Sys.iswindows()
+    using OrdinaryDiffEq: PIDController
+    trixi_include(@__MODULE__, joinpath(EXAMPLES_DIR, "elixir_euler_density_wave_extended.jl"), controller = PIDController(1.0,0.5,0.2))
+    l2, linf = analysis_callback(sol)
+    eval(:(@test_trixi_include(joinpath(EXAMPLES_DIR, "elixir_euler_density_wave_restart.jl"), controller = PIDController(1.0,0.5,0.2), 
```
Why do you need the call to `eval`?

sloede commented on test/runtests.jl:
```diff
@@ -42,8 +42,7 @@ const TRIXI_NTHREADS = clamp(Sys.CPU_THREADS, 2, 3)
         @test true
-
-        run(`$(Base.julia_cmd()) --threads=$TRIXI_NTHREADS --check-bounds=yes --code-coverage=none $(abspath("test_threaded.jl"))`)
+        run(`$(Base.julia_cmd()) --threads=$TRIXI_NTHREADS --check-bounds=yes $(abspath("test_threaded.jl"))`)
```
Why is this changed in this PR? This seems unrelated to the original purpose of the PR?!

ArseniyKholod replied:
It was a test discussed in trixi-students to enable coverage for threaded CI and MPI on Linux. It is seems to work without issues, I'll publish new PR for that.

sloede replied:
OK, thanks!

sloede commented on src/callbacks_step/save_restart_dg.jl:
```diff
@@ -327,4 +327,53 @@ function load_restart_file_on_root(mesh::Union{ParallelTreeMesh, ParallelP4estMe
+    if mpi_isroot()
+        save_restart_controller(integrator, controller, restart_callback)
+    end
+    if mpi_isparallel()
+        MPI.Barrier(mpi_comm())
+    end
```
Why do you need a barrier here? If it is really necessary, it should be documented why.

ArseniyKholod replied:
You're right, it's not necessary any more.

sloede commented on src/callbacks_step/save_restart_dg.jl:
```diff
@@ -327,4 +327,53 @@ function load_restart_file_on_root(mesh::Union{ParallelTreeMesh, ParallelP4estMe
     return u_ode
 end
+
+function save_restart_controller(mesh, integrator,
```
Why pass the `mesh` as an argument? It seems unused...

ArseniyKholod replied:
You are right

sloede reviewed (changes requested):
Thanks for getting the ball rolling with this! IMHO, a big no-no in the current approach is that it requires to add OrdinaryDiffEq.jl as a dependency to Trix.jl, which we have not done in the past (and for good reason).

Please check if there is a way around this. For example, when storing the controller data, it should be sufficient to just query the types for the existence of certain fields to figure out what to store. As for loading controller data, we should be able to do so without knowing the type beforehand, or do we?

I haven't done a full review yet, since I think rewriting this to work without OrdinaryDiffEq.jl as a dependency will probably touch a lot of the implementation.

ArseniyKholod commented on Project.toml:
```diff
@@ -19,6 +19,7 @@ MPI = "da04e1cc-30fd-572f-bb4f-1f8673147195"
 MuladdMacro = "46d2c3a1-f734-5fdb-9937-b9b9aeba4221"
 Octavian = "6fd5a793-0b7e-452c-907f-f8bfe9c57db4"
 OffsetArrays = "6fe1bfb0-de20-5000-8ca7-80f57d26f881"
+OrdinaryDiffEq = "1dea7af3-3e70-54e6-95c3-0bf5283fa5ed"
```
I'll try to go around it

ArseniyKholod replied:
Now dependency is removed

sloede commented on src/callbacks_step/save_restart_dg.jl:
```diff
@@ -327,4 +327,28 @@ function load_restart_file_on_root(mesh::Union{ParallelTreeMesh, ParallelP4estMe
+        # Open file (preserve existing content)
+        h5open(filename, "cw") do file
+            # Add context information as attributes
+            attributes(file)["qold"] = integrator.qold
```
Please prefix the values with something such that an unaware user has an idea of what this is all for. E.g., instead of
* `qold`
* `dtpropose`
* `controller_err`
use
* `time_integrator_qold`
* `time_integrator_dtpropose`
* `time_integrator_controller_err`

sloede commented on src/callbacks_step/save_restart_dg.jl:
```diff
@@ -327,4 +327,28 @@ function load_restart_file_on_root(mesh::Union{ParallelTreeMesh, ParallelP4estMe
+            # Ensure that `dtpropose` is written as a double precision scalar
+            attributes(file)["dtpropose"] = convert(Float64, integrator.dtpropose)
+            # For PIDController is necessary to save additional parameters
+            if :err in fieldnames(typeof(controller))
+                attributes(file)["controller_err"] = controller.err
+            end
```
Maybe a simpler way to implement this would be with [`hasproperty`](https://docs.julialang.org/en/v1/base/base/#Base.hasproperty)?
```suggestion
            if hasproperty(controller, :err)
                attributes(file)["controller_err"] = controller.err
            end
```

sloede commented on src/callbacks_step/save_restart.jl:
```diff
@@ -156,5 +160,32 @@ function load_restart_file(semi::AbstractSemidiscretization, restart_file)
+            integrator.reeval_fsal = true
+        end
+        # Load additional parameters for PIDController
+        if (:err in fieldnames(typeof(controller))) &&
+           ("controller_err" in keys(attributes(file)))
```
I wouldn't silently allow to use a controller that does not get its required values from the restart file (it should be there already, right?
```suggestion
        if hasproperty(controller, :err)
```

sloede commented on src/callbacks_step/save_restart.jl:
```diff
@@ -156,5 +160,32 @@ function load_restart_file(semi::AbstractSemidiscretization, restart_file)
+        # Load additional parameters for PIDController
+        if (:err in fieldnames(typeof(controller))) &&
+           ("controller_err" in keys(attributes(file)))
+            controller.err[1:3] = read(attributes(file)["controller_err"])
```
If you do not know that it will always be 3 values, you could also try to do
```suggestion
            controller.err[:] = read(attributes(file)["controller_err"])
```
This might have the benefit to throw an error if the number of values do not match

sloede commented on src/callbacks_step/save_restart.jl:
```diff
@@ -156,5 +160,32 @@ function load_restart_file(semi::AbstractSemidiscretization, restart_file)
+
+Load the context information for PID- and PIControllers saved in a `restart_file`.
+"""
+function load_controller!(integrator, restart_file::AbstractString)
```
As stated in a comment later, I'd suggest to make this function more descriptive by naming it something like "load_adaptive_time_integrator!"

sloede commented on src/callbacks_step/save_restart.jl:
```diff
@@ -156,5 +160,32 @@ function load_restart_file(semi::AbstractSemidiscretization, restart_file)
+"""
+    load_controller!(integrator, restart_file::AbstractString)
+
+Load the context information for PID- and PIControllers saved in a `restart_file`.
```
```suggestion
Load the context information for error-based time integrators saved in a `restart_file`.
```
Less technical and more informative to users who don't know any details

sloede commented on src/callbacks_step/save_restart.jl:
```diff
@@ -156,5 +160,32 @@ function load_restart_file(semi::AbstractSemidiscretization, restart_file)
+    # Read context information for controller
+    h5open(restart_file, "r") do file
+        # Ensure that necessary information was saved
+        if ("qold" in keys(attributes(file))) &&
+           ("dtpropose" in keys(attributes(file)))
```
Instead of silently ignoring the case where the attributes are not present, I would do first a check on all properties that you expect, i.e., something like "if qold or dtpropose or err are missing, show an error 'missing data in restart file: ...'" If someone does not want to load these values, they can just not call the load function. But if they _do_ use this function, they should be able to expect that everything works

sloede commented on src/callbacks_step/save_restart.jl:
```diff
@@ -156,5 +160,32 @@ function load_restart_file(semi::AbstractSemidiscretization, restart_file)
+           ("dtpropose" in keys(attributes(file)))
+            integrator.qold = read(attributes(file)["qold"])
+            integrator.dtpropose = read(attributes(file)["dtpropose"])
+            # Accept step to use dtpropose already in the first step
+            integrator.accept_step = true
+            # Reevaluate integrator.fsal_first on the first step
+            integrator.reeval_fsal = true
```
I think these two settings do not need to be in the `if` condition but should always be set

sloede commented on examples/tree_2d_dgsem/elixir_euler_density_wave_extended.jl:
Instead of adding a completely new elixir, would it be possible to modify an existing one (e.g., one that already uses adaptive time integrators) to test the writing of the additional restart data?

ArseniyKholod replied:
I modified `tree_2d_dgsem/elixir_advection_restart.jl` so, if one will pass adaptive time integrators `alg`  into `trixi_include`, it will test new functionality.
https://github.com/trixi-framework/Trixi.jl/pull/1565/files#diff-6f3d352433bd9e2d2fe073adb672ad8cce00af777b4062db6b4b5e8912ef4df6

sloede replied:
OK. Why did you change the existing values for tspan etc. though? I'm not saying you shouldn't, but usually it makes it easier for the reviewer (and yourself) if you keep as much as it is, and do not change existing setups without good reason.

ArseniyKholod replied:
I.e. with `RDPK3SpFSAL49()` otherwise it takes only 6 steps for restart setup. I thought it might be not enough to identify some discrepancy. I wanted to be sure, should I do not do it?

sloede replied:
No, that sounds reasonable 👍

sloede commented on examples/tree_2d_dgsem/elixir_euler_density_wave_restart.jl:
```diff
@@ -0,0 +1,46 @@
+load_controller!(integrator, restart_filename)
+
+# Get the last time index and work with that.
+integrator.iter = load_timestep(restart_filename)
+integrator.stats.naccept = integrator.iter
```
Maybe in the next PR, we should also add a function `load_timestep!(integrator, restart_filename` to make this part easier 🙂

ArseniyKholod replied:
For this purpose, I created a new PR #1614

ArseniyKholod commented on test/test_mpi_tree.jl:
```diff
@@ -189,6 +189,20 @@ CI_ON_WINDOWS = (get(ENV, "GITHUB_ACTIONS", false) == "true") && Sys.iswindows()
+  @trixi_testset "elixir_euler_density_wave_restart.jl" begin
+    trixi_include(@__MODULE__, joinpath(EXAMPLES_DIR, "elixir_euler_density_wave_extended.jl"));
+    l2_expected, linf_expected = analysis_callback(sol)
+    Trixi.mpi_isroot() && println("═"^100)
+    Trixi.mpi_isroot() && println(joinpath(EXAMPLES_DIR, "elixir_euler_density_wave_restart.jl"))
```
Thank you for the review! It's a part of actual test, since I don't call `@test_trixi_include`, the running file name in test won't be printed without these lines and it will be complicated in case of error to identify the problematic elixir. The same was implemented here:
https://github.com/trixi-framework/Trixi.jl/blob/e22b11e5859b5168e24dabb4cb34a6a35fce3e66/test/test_mpi_p4est_2d.jl#L22
Do you think it is unnecessary?

sloede replied:
OK, thanks for the explanation. No, I think it that case it is a good idea to have it 👍

ArseniyKholod commented on src/callbacks_step/save_restart_dg.jl:
```diff
@@ -327,4 +327,28 @@ function load_restart_file_on_root(mesh::Union{ParallelTreeMesh, ParallelP4estMe
+            # Add context information as attributes
+            attributes(file)["qold"] = integrator.qold
+            # Ensure that `dtpropose` is written as a double precision scalar
+            attributes(file)["dtpropose"] = convert(Float64, integrator.dtpropose)
```
It should be removed, sorry

ArseniyKholod commented on src/callbacks_step/save_restart.jl:
```diff
@@ -105,6 +105,10 @@ function (restart_callback::SaveRestartCallback)(integrator)
         end
 
         save_restart_file(u_ode, t, dt, iter, semi, restart_callback)
+        if integrator.opts.adaptive
```
As I understand, every defined integration scheme in `src/time_integration` has always `integrator.opts.adaptive = false` and hasn't field `integrator.opts.controller`. So it will be always `if false` and no controller info will be saved, I think it's a right behavior in tis case

sloede replied:
OK, I just wanted to confirm that there are no "type whatever has no field adaptive" errors. Our own schemes are not adaptive, so yes, this is the correct behavior 👍

ArseniyKholod replied:
I didn't get an error, i.e. with `Trixi.SimpleSSPRK33()`

ArseniyKholod commented:
Michael, @sloede, may be you have some quick ideas, why this error could appear? Something wrong with allocations. But error appeared first time when I added only some comments, without changing a code, strange, I'm completely at a loss.
https://github.com/trixi-framework/Trixi.jl/actions/runs/5977210124/job/16216692622?pr=1565#step:7:3322
The same error I got in my new PR (https://github.com/trixi-framework/Trixi.jl/pull/1614), these PRs seem to have nothing in common.
https://github.com/trixi-framework/Trixi.jl/actions/runs/5977692959/job/16218213885?pr=1614#step:7:3322
One more failing test: https://github.com/trixi-framework/Trixi.jl/actions/runs/5977692959/job/16218213885?pr=1614#step:7:8346

ranocha commented on src/callbacks_step/save_restart.jl:
```diff
@@ -156,5 +161,34 @@ function load_restart_file(semi::AbstractSemidiscretization, restart_file)
+"""
+    load_adaptive_time_integrator!(integrator, restart_file::AbstractString)
+
+Load the context information for error-based time integrators saved in a `restart_file`.
```
```suggestion
Load the context information for time integrators with error-based step size control
saved in a `restart_file`.
```

ranocha commented on src/callbacks_step/save_restart.jl:
```diff
@@ -156,5 +161,34 @@ function load_restart_file(semi::AbstractSemidiscretization, restart_file)
+        integrator.accept_step = true
+        # Reevaluate integrator.fsal_first on the first step
+        integrator.reeval_fsal = true
+        # Load additional parameters for PIDController
+        if hasproperty(controller, :err)
+            controller.err[:] = read(attributes(file)["time_integrator_controller_err"])
+        end
```
This has to be behind some additional checks which controller is used. We default to `PIDController`s in many cases, but other algorithms use a `PIController`. See https://docs.sciml.ai/DiffEqDocs/stable/extras/timestepping/

ArseniyKholod replied:
I'm not sure, that I understood correctly. Do you mean to check before explicitly whether controller is `PIDController` or `PIController`? If yes, it requires a dependency on `OrdinaryDifffEq`, where these types are defined. This package is huge and Michael said better not to depend on it for efficiency.
So check `hasproperty(controller, :err)` is exactly to distinguish between PID and PI without loading OrdinaryDiffEq.
So I'm not sure what I need to change. Now it works both for `PIDController` and `PIController`

ranocha replied:
Please explain this in a comment

ArseniyKholod replied:
Okey, I tested with `SSPRK43` in [one of the tests](https://github.com/trixi-framework/Trixi.jl/pull/1565/files#diff-43aaf8164d76b0a0bff6056e14898a55083136429d17b91aad6d67bd85e2473b)

ranocha commented on src/callbacks_step/save_restart_dg.jl:
```diff
@@ -327,4 +327,29 @@ function load_restart_file_on_root(mesh::Union{ParallelTreeMesh, ParallelP4estMe
+            attributes(file)["time_integrator_qold"] = integrator.qold
+            # Ensure that `dtpropose` is written as a double precision scalar
+            attributes(file)["time_integrator_dtpropose"] = integrator.dtpropose
+            # For PIDController is necessary to save additional parameters
+            if hasproperty(controller, :err)
+                attributes(file)["time_integrator_controller_err"] = controller.err
+            end
```
Here too

sloede reviewed (approved):
LGTM! It would be good to also get a review from @ranocha

sloede commented on src/callbacks_step/save_restart.jl:
```diff
@@ -156,5 +160,32 @@ function load_restart_file(semi::AbstractSemidiscretization, restart_file)
+            # Accept step to use dtpropose already in the first step
+            integrator.accept_step = true
+            # Reevaluate integrator.fsal_first on the first step
+            integrator.reeval_fsal = true
```
@ranocha it would be great if you could confirm that this makes sense.

ranocha reviewed (commented):
Did you add a test using a `PIDController`?

ArseniyKholod commented:
> Did you add a test using a `PIDController`?

Yes, in [test_mpi_tree.jl](https://github.com/trixi-framework/Trixi.jl/pull/1565/files#diff-0d913e7961860bd3a9c595d68dbd7a7b09ea17758afc8c814b9e45a454a04796) and code part, that is used only by `PIDController`, is covered by coverage test
